assignment5.cpp

```diff
@@ -0,0 +1,219 @@
+    
+    for (map<char, vector<char> >::iterator it = adjacency_matrix.begin(); it!=adjacency_matrix.end(); ++it) {
+        vector<char> adj_chars = it -> second;
+        for (int i = 0; i < adj_chars.size(); i++) {
+            if (!in_degrees.count(adj_chars[i])) {
+                in_degrees.insert(make_pair(adj_chars[i], 1));
+            } else {
+                in_degrees[adj_chars[i]]++;
+            }
+        }
+    }
+    
+    return in_degrees;
+}
+
+/* visits node in the dag */
+void visit(char char_to_visit, map<char, int> *in_degrees, set<char> *visited, map<char, vector<char> > adj_matrix, queue<char> *roots_of_dag) {
+    vector<char> adj_chars = adj_matrix[char_to_visit];
+    
+    visited->insert(char_to_visit);
+    
+    for (int i = 0; i < adj_chars.size(); i++) {
+        if (!visited->count(adj_chars[i])) {
+            (*in_degrees)[adj_chars[i]]--;
+            if ((*in_degrees)[adj_chars[i]] == 0) {
+                roots_of_dag->push(adj_chars[i]);
+            }
+        }
+    }
+}
+
+/* does the topological sorting and returns the ordered alphabet */
+vector<char> topological_sort(map<char, vector<char> > adjacency_matrix, map<char, int> in_degrees, vector<string> dictionary) {
+    vector<char> ordered_alphabet;
+    set<char> visited;
+    queue<char> roots_of_dag;
+    
+    /* adds all chars with 0 in degree to the queue */
+    for (int i = 0; i < dictionary.size(); i++) {
+        for (int j = 0; j < dictionary[i].size(); j++) {
+            if (!in_degrees.count(dictionary[i][j])) {
+                roots_of_dag.push(dictionary[i][j]);
+            }
+        }
+    }
+    
+    while (!roots_of_dag.empty()) {
+        char curr_node = roots_of_dag.front();
+        roots_of_dag.pop();
+        if (!visited.count(curr_node)) {
+            ordered_alphabet.push_back(curr_node);
+            visit(curr_node, &in_degrees, &visited, adjacency_matrix, &roots_of_dag);
+        }
+    }
+    
+
+    
+    return ordered_alphabet;
+}
+
+/* returns ordered alphabet given a dictionary */
+vector<char> alphabet(vector<string> dictionary) {
+    vector<char> ordered_alphabet;
+    map<char, vector<char> > adjacency_matrix;
+    
+    /* build adjacency matrix for graph */
+    adjacency_matrix = build_adjacency_matrix(dictionary);
+    
+    /* find in degrees for topological sort */
+    map<char, int> in_degrees = find_in_degrees(adjacency_matrix);
+    
+    /* topological sorting using dfs */
+    ordered_alphabet = topological_sort(adjacency_matrix, in_degrees, dictionary);
+    
+    return ordered_alphabet;
+}
+
+void given_test() {
+    vector<string> dictionary;
+    string a = "ART";
+    dictionary.push_back(a);
+    a = "RAT";
+    dictionary.push_back(a);
+    a = "CAT";
+    dictionary.push_back(a);
+    a = "CAR";
+    dictionary.push_back(a);
+    
+    vector<char> result = alphabet(dictionary);
+    
+    for (int i = 0; i < result.size(); i++) {
+        cout<<result[i]<<" ";
```

Comment:
You can write what it should print as a comment, so you can refer to it later (and so I can see it and maybe have comments about it...).